bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/Stream2JSONInputStream.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.rest;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.apache.commons.io.IOUtils;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * This {@link InputStream} will stream {@link Stream}s as JSON one item at a time. This will reduce memory usage when
+ * streaming large collections through the REST interface. The input stream creates one JSON representation at a time
+ * from the top level elements of the stream. For best performance a flattened stream should be provided. Otherwise a
+ * nested collections JSON representation will be fully transformed into memory.
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+public class Stream2JSONInputStream extends InputStream {
+
+    private final Iterator<String> iterator;
+
+    private InputStream jsonElementStream;
+
+    private boolean firstIteratorElement;
+
+    private Gson GSON = new GsonBuilder().create();
+
+    /**
+     * Creates a new {@link Stream2JSONInputStream} backed by the given {@link Stream} source.
+     *
+     * @param source the {@link Stream} backing this input stream. Must not be null.
+     *
+     * @throws IllegalArgumentException in case the source is null.
+     */
+    public Stream2JSONInputStream(Stream<?> source) {
+        if (source == null) {
+            throw new IllegalArgumentException("The source must not be null!");
+        }
+
+        iterator = source.map(e -> GSON.toJson(e)).iterator();
+        jsonElementStream = IOUtils.toInputStream("");
+        firstIteratorElement = true;
+    }
+
+    @Override
+    public int read() throws IOException {
+        int result = jsonElementStream.read();
+
+        if (result == -1) { // the current JSON element was completely streamed
+            if (finished()) { // we are done streaming the collection
+                return -1;
+            }
+
+            fillBuffer(); // get the next element into the jsonInputStream
+            result = jsonElementStream.read();
+        }
+
+        return result;
+    }
+
+    @Override
+    public void close() throws IOException {
+        closeJSONStream();
+        super.close();
```

Comment:
`InputStream.close()` doesn't contain any implementation so you can drop that call (but it doesn't matter).
But I am pretty sure that the base class close will not get an implementation some time because it would break all previously working subclasses.